File: 프로그래머스/3/81303. 표 편집/표 편집.py
def solution(n, k, cmd):

    answer = ['O'] * n
    # 스택을 사용하여 삭제된 행을 저장
    removedStack = []
    # 현재 행
    curr = k
    # 이전, 다음 인덱스
    prev = [i - 1 for i in range(n)]
    next = [i + 1 for i in range(n)]
    # 마지막 행 이후의 값은 존재하지 않음
    next[-1] = -1

    for command in cmd:
        # 삭제
        if command[0] == "C":
            # 현재 행 삭제
            removedStack.append((curr, prev[curr], next[curr]))
            answer[curr] = 'X'

            # 연결 끊기
            if prev[curr] != -1:
                next[prev[curr]] = next[curr]
            if next[curr] != -1:
                prev[next[curr]] = prev[curr]

            # 현재 행 업데이트
            if next[curr] != -1:
                curr = next[curr]
            else:
                curr = prev[curr]

        # 복구
        elif command[0] == "Z":
            r, p, n = removedStack.pop()
            answer[r] = 'O'

            if p != -1:
                next[p] = r
            if n != -1:
                prev[n] = r

        # 선택
        elif command[0] == "U":
            X = int(command.split()[1])
            for _ in range(X):
                curr = prev[curr]

        elif command[0] == "D":
            X = int(command.split()[1])
            for _ in range(X):
                curr = next[curr]

    return ''.join(answer)

File: 프로그래머스/3/81303. 표 편집/표 편집.js
function solution(n, k, cmd) {
    let answer = Array(n).fill('O');
    let removedStack = [];
    let curr = k;
    let prev = Array.from({length: n}, (_, i) => i - 1);
    let next = Array.from({length: n}, (_, i) => i + 1);
    next[n - 1] = -1; // 마지막 행 이후의 값은 존재하지 않음

    cmd.forEach(command => {
        if (command.startsWith("C")) {
            // 현재 행 삭제
            removedStack.push([curr, prev[curr], next[curr]]);
            answer[curr] = 'X';

            // 연결 끊기
            if (prev[curr] !== -1) {
                next[prev[curr]] = next[curr];
            }
            if (next[curr] !== -1) {
                prev[next[curr]] = prev[curr];
            }

            // 현재 행 업데이트
            if (next[curr] !== -1) {
                curr = next[curr];
            } else {
                curr = prev[curr];
            }

        } else if (command.startsWith("Z")) {
            // 가장 최근에 삭제된 행 복구
            const [r, p, n] = removedStack.pop();
            answer[r] = 'O';

            if (p !== -1) {
                next[p] = r;
            }
            if (n !== -1) {
                prev[n] = r;
            }

        } else {
            const [direction, X] = command.split(" ");
            const steps = parseInt(X);
            if (direction === "U") {
                for (let i = 0; i < steps; i++) {
                    curr = prev[curr];
                }
            } else if (direction === "D") {
                for (let i = 0; i < steps; i++) {
                    curr = next[curr];
                }
            }
        }
    });

    return answer.join('');
};
